src/openrct2/ride/gentle/Maze.cpp

```diff
@@ -181,3 +201,223 @@ TRACK_PAINT_FUNCTION get_track_paint_function_maze(int32_t trackType, int32_t di
+
+void MazePathfindingEntry::SetXTile(uint16_t xTile)
+{
+    SetField<uint32_t>(Data, xTile, 11, 5);
+}
+
+uint16_t MazePathfindingEntry::GetXTile() const
+{
+    return GetField(Data, 11, 5);
+}
+
+void MazePathfindingEntry::SetVisitedFlags(uint8_t visitedFlags)
+{
+    SetField<uint32_t>(Data, visitedFlags, 4, 16);
+}
+
+void MazePathfindingEntry::AddVisitedFlags(Direction visitedDirection)
+{
+    SetVisitedFlags(GetVisited() | (1 << GetField(visitedDirection, 2)));
+}
+
+void MazePathfindingEntry::SetCompletlyVisited(bool completlyVisited)
+{
+    SetField<uint32_t>(Data, completlyVisited, 1, 20);
+}
+
+void MazePathfindingEntry::SetYTile(uint16_t yTile)
+{
+    SetField<uint32_t>(Data, yTile, 11, 21);
+}
+
+uint16_t MazePathfindingEntry::GetYTile() const
+{
+    return GetField(Data, 11, 21);
+}
+
+void MazePathfindingEntry::Reset()
+{
+    Data = 0x0;
+}
+
+void MazePathfindingEntry::Free()
+{
+    Data = ~0x0u;
+}
+
+void MazePathfindingEntry::Set(const CoordsXY& coords, uint8_t subTileIndex, Direction origin)
+{
+    assert(coords.x > 0 && coords.y > 0);
+    TileCoordsXY tileCoords(coords);
+
+    Reset();
+    SetSubTileIndex(subTileIndex);
+    SetOrigin(origin);
+    SetXTile(tileCoords.x);
+    SetYTile(tileCoords.y);
+}
+
+std::pair<CoordsXY, uint8_t> MazePathfindingEntry::GetCoords() const
+{
+    TileCoordsXY tileCoords(GetXTile(), GetYTile());
+
+    return std::pair(tileCoords.ToCoordsXY(), GetSubTileIndex());
+}
+
+Direction MazePathfindingEntry::GetOrigin() const
+{
+    return GetField(Data, 2, 2);
+}
+
+void MazePathfindingEntry::MarkVisited(Direction source, uint8_t openEdgesCount)
+{
+    AddVisitedFlags(source);
+
+    if (bitcount(GetVisited()) >= openEdgesCount - 1)
+        SetCompletlyVisited(true);
+}
+
+uint8_t MazePathfindingEntry::GetVisited() const
+{
+    return GetField(Data, 4, 16);
+}
+
+bool MazePathfindingEntry::IsCompletlyVisited() const
+{
+    return GetField(Data, 1, 20);
+}
+
+bool MazePathfindingEntry::MatchCoords(const CoordsXY& coords, uint8_t subTileIndex) const
+{
+    TileCoordsXY tileCoords(coords);
+
+    return (GetXTile() == tileCoords.x) && (GetYTile() == tileCoords.y) && (GetSubTileIndex() == subTileIndex);
+}
+
+bool MazePathfindingEntry::MatchCoords(const MazePathfindingEntry& e) const
+{
+    return (GetXTile() == e.GetXTile()) && (GetYTile() == e.GetYTile()) && (GetSubTileIndex() == e.GetSubTileIndex());
+}
+
+void MazePathfindingHistory::Initialize()
+{
+    for (auto& item : Stack)
+        item.Reset();
+}
+
+void MazePathfindingHistory::Free()
+{
+    for (auto& item : Stack)
+        item.Free();
+}
+
+std::pair<MazePathfindingEntry&, uint8_t> MazePathfindingHistory::MeetIntersection(
+    const CoordsXY& coords, uint8_t subTileIndex, Direction source)
+{
+    uint8_t i;
+
+    for (i = 0; i < Stack.size(); ++i)
+    {
+        if (Stack[i].MatchCoords(coords, subTileIndex))
+            break;
+    }
+
+    for (uint8_t j = std::min((uint8_t)(Stack.size() - 1), i); j > 0; --j)
```

Comment:
```suggestion
    for (uint8_t j = std::min(Stack.size() - 1, i); j > 0; --j)
```
I think you can let it implicitly cast safely

Reply:
This is also my opinion but not msvc one.